Mathlib/Tactic/Linarith/Parsing.lean

```diff
@@ -180,7 +180,8 @@ and forces some functions that call it into `MetaM` as well.
 -/
 
 partial def linearFormOfExpr (red : TransparencyMode) (m : ExprMap) (e : Expr) :
-    MetaM (ExprMap × Sum) :=
+    MetaM (ExprMap × Sum) := do
+  let e ← whnfR e
```

Comment:
Isn't `consumeMData` the usual way to ignore metadata? In particular, I'm anxious about whnf doing more things (eg slower things) than just discarding metadata

Reply:
My argument for `whnfR` is that `~q` matching [uses `withReducible`](https://github.com/leanprover-community/quote4/blob/e1d2994e0acdee2f0c03c9d84d28a5df34aa0020/Qq/Match.lean#L130) by default, and is used throughout norm_num and positivity; so I think users should `linarith` to behave similarly.

`whnf` doing unfolding should be fine here because the user will never see the effect of the unfolding, since `linarith` is terminal.

Reply:
> (Indeed, I have a branch in the works to rewrite this with Qq for speed)

Is this PR relevant then?

> `whnf` doing unfolding should be fine here because the user will never see the effect of the unfolding, since `linarith` is terminal.

What if a term unfolds unexpectedly to more arithmetic that `linarith` can process? What if there is a predicate in context that unfolds to linear arithmetic? Do we know that these wouldn't cause `linarith` to fail?

Are we happy with Qq using reducible isDefEq? I know I'm still unsettled by that. I think there's also a difference between matching things using isDefEq and matching things structurally, like the function currently does.

Reply:
> Is this PR relevant then?

Maybe rather I should say "if we are fixing a metadata bug, and you have a PR in the pipeline that will change the strategy, could we please either use `consumeMData` or `whnfCore` here to preserve `linarith`'s current strategy, so that we can merge this without needing any discussion"?